art/attacks/evasion/graphite/graphite_whitebox_pytorch.py

```diff
@@ -0,0 +1,565 @@
+        x_adv_copy = np.zeros((x_adv.shape[0], self.net_size[1], self.net_size[0], x_adv.shape[1]))
+        x_copy_trans = np.transpose(x, (0, 2, 3, 1))
+        x_adv_copy_trans = np.transpose(x_adv, (0, 2, 3, 1))
+        for i in range(x_copy.shape[0]):
+            x_copy[i] = convert_to_network(x_copy_trans[i], self.net_size, clip_min, clip_max)
+            x_adv_copy[i] = convert_to_network(x_adv_copy_trans[i], self.net_size, clip_min, clip_max)
+
+        if self.estimator.channels_first:
+            x_copy = np.transpose(x_copy, (0, 3, 1, 2))
+            x_adv_copy = np.transpose(x_adv_copy, (0, 3, 1, 2))
+
+        logger.info(
+            "Success rate of GRAPHITE white-box attack: %.2f%%",
+            100
+            * compute_success(
+                self.estimator,
+                x_copy.astype(np.float32),
+                y,
+                x_adv_copy.astype(np.float32),
+                self.targeted,
+                batch_size=self.batch_size,
+            ),
+        )
+
+        if not self.estimator.channels_first:
+            x_adv = np.transpose(x_adv, (0, 2, 3, 1))
+
+        return x_adv
+
+    def _eval(
+        self,
+        x: "torch.Tensor",
+        x_adv: "torch.Tensor",
+        mask: "torch.Tensor",
+        target_label: int,
+        y_onehot: "torch.Tensor",
+        xforms: List[Tuple[float, float, float, int, float, float, float, float, float]],
+        pts: Optional[np.ndarray],
+        clip_min: float,
+        clip_max: float,
+    ) -> float:
+        """
+        Compute transform-robustness.
+        :param x: Original image.
+        :param x_adv: Attacked image.
+        :param mask: The mask.
+        :param target_label: The target label.
+        :param y_onehot: The target label in one hot form.
+        :param xforms: Ths list of transformation parameters.
+        :param pts: Optional. A set of points that will set the crop size in the perspective transform.
+        :param clip_min: Minimum value of an example.
+        :param clip_max: Maximum value of an example.
+        :return: Transform-robustness of the attack.
+        """
+        import torch  # lgtm [py/repeated-import]
+
+        successes = 0
+        for xform in xforms:
+            with torch.no_grad():
+                if len(x_adv.shape) == 3:
+                    x_adv = x_adv.unsqueeze(0)
+                transformed_x_adv = transform_wb(x, x_adv, mask, xform, pts, self.net_size, clip_min, clip_max)
+                logits, _ = self.estimator._predict_framework(  # pylint: disable=W0212
+                    transformed_x_adv.cuda(), y_onehot
+                )
+                success = int(logits.argmax(dim=1).detach().cpu().numpy()[0] == target_label)
+                successes += success
+        return successes / len(xforms)
+
+    def _perturb(
+        self,
+        x: np.ndarray,
+        y: np.ndarray,
+        mask: np.ndarray,
+        pts: Optional[np.ndarray],
```

Comment:
Should assign it with a default value of None and move to the end of the arg list.